src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/ProtobufSerializer.cs

```diff
@@ -42,63 +42,11 @@ internal static int WriteTagAndLength(byte[] buffer, int writePosition, int cont
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
     internal static void WriteReservedLength(byte[] buffer, int writePosition, int length)
     {
-        int byteLength = 0;
-        int? firstByte = null;
-        int? secondByte = null;
-        int? thirdByte = null;
-        int? fourthByte = null;
-
-        do
-        {
-            switch (byteLength)
-            {
-                case 0:
-                    firstByte = length & 0x7F;
-                    break;
-                case 1:
-                    secondByte = length & 0x7F;
-                    break;
-                case 2:
-                    thirdByte = length & 0x7F;
-                    break;
-                case 3:
-                    fourthByte = length & 0x7F;
-                    break;
-            }
-
-            length >>= 7;
-            byteLength++;
-        }
-        while (length > 0);
-
-        if (fourthByte.HasValue)
-        {
-            buffer[writePosition++] = (byte)(firstByte!.Value | 0x80);
-            buffer[writePosition++] = (byte)(secondByte!.Value | 0x80);
-            buffer[writePosition++] = (byte)(thirdByte!.Value | 0x80);
-            buffer[writePosition++] = (byte)fourthByte!.Value;
-        }
-        else if (thirdByte.HasValue)
-        {
-            buffer[writePosition++] = (byte)(firstByte!.Value | 0x80);
-            buffer[writePosition++] = (byte)(secondByte!.Value | 0x80);
-            buffer[writePosition++] = (byte)(thirdByte!.Value | 0x80);
-            buffer[writePosition++] = 0;
-        }
-        else if (secondByte.HasValue)
-        {
-            buffer[writePosition++] = (byte)(firstByte!.Value | 0x80);
-            buffer[writePosition++] = (byte)(secondByte!.Value | 0x80);
-            buffer[writePosition++] = 0x80;
-            buffer[writePosition++] = 0;
-        }
-        else
-        {
-            buffer[writePosition++] = (byte)(firstByte!.Value | 0x80);
-            buffer[writePosition++] = 0x80;
-            buffer[writePosition++] = 0x80;
-            buffer[writePosition++] = 0;
-        }
+        var slice = buffer.AsSpan(writePosition, 4);
+        slice[0] = (byte)(((length >> 0) & 0x7F) | 0x80);
```

Comment:
I think that this line can have one more micro optimization. I am not sure how smart is the compiler:

```suggestion
        slice[0] = (byte)((length & 0x7F) | 0x80);
```

Reply:
Re-ran the microbenchmarks I wrote with this change and no obvious impact so all good.

Reply:
Writting high performance C# is mostly about using spans and memory instead of arrays, even if backed by arrays.

This method has no need for an array. In fact, the first thing it does is creating a span over it.

This

```csharp
internal static void WriteReservedLength(Span<byte> buffer, int length)
```

would decouple the method from the underlying memory store for the buffer.

Reply:
Happy for @paulomorgado to do the follow-up.